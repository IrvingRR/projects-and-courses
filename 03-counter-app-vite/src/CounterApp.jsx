import { Fragment, useState } from 'react';
import PropTypes from 'prop-types';

const CounterApp = ({ value }) => {

    const [ counter, setCounter ] = useState(value);

    const handleIncrement = () => setCounter(counter + 1);
    const handleSubtract = () => setCounter(counter - 1);
    const handleReset = () => setCounter(value);

    return (
        <Fragment>
            <h1>Counter</h1>
            <h2>{ counter }</h2>
            <button aria-label='btn-increment' onClick={ handleIncrement }>+1</button>
            <button aria-label='btn-decrement' onClick={ handleSubtract }>-1</button>
            <button aria-label='btn-reset' onClick={ handleReset }>Reset</button>
        </Fragment>
    );
};

CounterApp.propTypes = {
    value: PropTypes.number
};

CounterApp.defaultProps = {
    value: 0
};

export default CounterApp;