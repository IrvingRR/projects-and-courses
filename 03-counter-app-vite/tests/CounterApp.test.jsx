import { render, screen, fireEvent } from '@testing-library/react';
import CounterApp from '../src/CounterApp';

describe('Tests in <CounterApp/>', () => {


    test('Should to do match with the snapshot', () => {
        const { container } = render(<CounterApp/>);
        expect(container).toMatchSnapshot();
    });

    test('Should to show ti initial value 100', () => {
        render(<CounterApp value={100}/>);
        const value = screen.getByRole('heading', { level: 2 });

        expect(value.innerHTML).toContain('100'); // First alternative
        expect(screen.getByText(100)).toBeTruthy(); // Second alternative

    });

    test('Should to increment with the button +1', () => {
         render(<CounterApp/>);

        const value = screen.getByText('0');
        const button = screen.getByText('+1');
        fireEvent.click(button);

        // expect(screen.getByText('1')).toBeTruthy();
        expect(value.innerHTML).toBe('1');

    });

    test('Should to decrement with the button -1', () => {
         render(<CounterApp value={10}/>);

        const value = screen.getByText('10');
        const button = screen.getByText('-1');
        fireEvent.click(button);

        // expect(screen.getByText('1')).toBeTruthy();
        expect(value.innerHTML).toBe('9');

    });

    test('Should to reset to the initial value when the reset button is clicked', () => {
        render(<CounterApp/>);
        const buttonIncrement = screen.getByText('+1');
        const buttonReset = screen.getByText('Reset');

        // Other alternative to get one specific element
        const btnIncrement = screen.getByRole('button', { name: 'btn-increment' });
        const btnReset = screen.getByRole('button', { name: 'btn-reset' });

        fireEvent.click(btnIncrement);
        fireEvent.click(btnIncrement);
        fireEvent.click(btnIncrement);
        fireEvent.click(btnReset);


        expect(screen.getByText('0')).toBeTruthy();
    });

});