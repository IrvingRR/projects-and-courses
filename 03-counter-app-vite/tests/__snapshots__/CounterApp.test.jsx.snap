// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <CounterApp/> Should to do match with the snapshot 1`] = `
<div>
  <h1>
    Counter
  </h1>
  <h2>
    0
  </h2>
  <button
    aria-label="btn-increment"
  >
    +1
  </button>
  <button
    aria-label="btn-decrement"
  >
    -1
  </button>
  <button
    aria-label="btn-reset"
  >
    Reset
  </button>
</div>
`;
